Add multiple statuses on proposals

alecslupu opened the pull request:
#### :tophat: What? Why?
*Please describe your pull request.*

#### :pushpin: Related Issues
*Link your PR to an issue*
- Related to [DDP31]
- Meta Proposal: [Expand participatory process status types](https://meta.decidim.org/processes/roadmap/f/122/proposals/15920)

#### Testing
1. Reacreate  your DB 
2. Visit proposals's index page in admin panel 
3. See there is a new button named "States"
4. Edit and Add various states 
5. Go to Proposals answer page and add and answer, including new States. 
6. Publish the answer, and see is available in interface
7. Also, you can use Proposal templates 

### :camera: Screenshots
*Please add screenshots of the changes you are proposing*

![image](https://github.com/tremend-cofe/decidim/assets/105683/4ec2e344-6a03-456f-b33f-aca1155a00d4)
![image](https://github.com/tremend-cofe/decidim/assets/105683/010751ea-8d04-4365-ab0f-5a2b2a6a80a5)
![image](https://github.com/tremend-cofe/decidim/assets/105683/66e75fc6-5606-4bd1-aee8-04f37d1c3e8f)
![image](https://github.com/tremend-cofe/decidim/assets/105683/bf1725fc-2638-4fdd-a1b8-77c6570f3684)
![image](https://github.com/decidim/decidim/assets/105683/7140473b-8e6f-41e7-8a68-ebe3e8de0de2)


:hearts: Thank you!

andreslucena commented:
Can you link to the metadecidim proposal 🙏🏽 ?

alecslupu commented on decidim-proposals/app/controllers/concerns/decidim/proposals/admin/filterable.rb:
```diff
@@ -55,13 +53,21 @@ def filters_with_values
           end
 
+          def translated_proposal_state_id_eq(state_id)
+            translated_attribute(ProposalState.find_by(id: state_id)&.title)
```
```suggestion
            translated_attribute(ProposalState.find_by(component: current_component, id: state_id)&.title)
```

andreslucena replied:
Should we apply this auto suggestion @alecslupu ?

alecslupu replied:
Is ok, we can skip it.

andreslucena commented:
We just give it a look with @carolromero and there are some things that we want to change (or at least to know better the limitations or implications on why things are like that):

1. `token` field: we think is too technical and if there's some other way of having this internally that'd be better (like generating it from the `state.title.to_sym` or something like that)
2. `description` field: is it used? if not lets drop it, as it doesn't make sense adding fields "just in case" 
3. `system` checkbox: what's the usage of this? 
4. `default` checkbox: what's the usage of this? 
5. The original need is to be able to customize the three States in proposals: "Evaluating"/"Accepted"/"Rejected". Can we remove "Withdrawn" and "Not answered" from this page? I know that you've mentioned it me in a past meeting, but talking about this with @carolromero just seems weird to be able to override these from the admin as they're "special" statuses. We couldn't foresee any case where we wanted to override these two, and on those cases you could always make a code override of course
6. "Answerable" and "Notifiable": all the status should be possible to select and answers and also generate notification (except the special cases of "Withdrawn" and "Not answered", that I understand is why these options exist)
7. "Gamified": Is that used only for the "Accepted proposals" badge? Can we left the badge as is for the moment (using the Accepted status if it exist)?
8. "Css class" we'd prefer to be able to select the color, as the use case for this one is the "Finished" state in Metadecidim with the purple color. We may need some logic to generate the contrast between the chosen color and the background. This should be a color picker as other colors fields that we have (in Admin -> Settings -> Appearance).

With all these changes the form should be much more simpler to grok and not so scary on a first look. This is what we had in mind:

![Mockup of the new "Edit state" form](https://github.com/decidim/decidim/assets/717367/706f1d95-e22e-4b72-b9fd-b5f9333ffdba)

alecslupu commented:
> We just give it a look with @carolromero and there are some things that we want to change (or at least to know better the limitations or implications on why things are like that):
> 
>     1. `token` field: we think is too technical and if there's some other way of having this internally that'd be better (like generating it from the `state.title.to_sym` or something like that)
> 
>     2. `description` field: is it used? if not lets drop it, as it doesn't make sense adding fields "just in case"
> 
>     3. `system` checkbox: what's the usage of this?
> 
>     4. `default` checkbox: what's the usage of this?
> 
>     5. The original need is to be able to customize the three States in proposals: "Evaluating"/"Accepted"/"Rejected". Can we remove "Withdrawn" and "Not answered" from this page? I know that you've mentioned it me in a past meeting, but talking about this with @carolromero just seems weird to be able to override these from the admin as they're "special" statuses. We couldn't foresee any case where we wanted to override these two, and on those cases you could always make a code override of course
> 
>     6. "Answerable" and "Notifiable": all the status should be possible to select and answers and also generate notification (except the special cases of "Withdrawn" and "Not answered", that I understand is why these options exist)
> 
>     7. "Gamified": Is that used only for the "Accepted proposals" badge? Can we left the badge as is for the moment (using the Accepted status if it exist)?
> 
>     8. "Css class" we'd prefer to be able to select the color, as the use case for this one is the "Finished" state in Metadecidim with the purple color. We may need some logic to generate the contrast between the chosen color and the background. This should be a color picker as other colors fields that we have (in Admin -> Settings -> Appearance).
> 
> 
> With all these changes the form should be much more simpler to grok and not so scary on a first look. This is what we had in mind:
> 
> ![Mockup of the new "Edit state" form](https://private-user-images.githubusercontent.com/717367/294908337-706f1d95-e22e-4b72-b9fd-b5f9333ffdba.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3MjQwMDQsIm5iZiI6MTcwNDcyMzcwNCwicGF0aCI6Ii83MTczNjcvMjk0OTA4MzM3LTcwNmYxZDk1LWUyMmUtNGI3Mi1iOWZkLWI1ZjkzMzNmZmRiYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDEwOFQxNDIxNDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01NDlkYzkyZTc5YzRiYjQ2MDgwNDY5YTQzOGJkNmQwYzE0MGFjMzc5OGMzNGNiZGY3NzIyY2U4ZjhmY2ZhZTRiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.B09cb9t658RetEBJdPqVWPPI_7lufFf2L2x0Z9dgt_o)

hello @andreslucena , please find below my feedback on your inquiry. At least this is what i had in mind when i have developed the module. 

1. token, is a string that an user can add. This token string is usable in filters. I have chosen not to generate it automatically, as sometimes, the state could be "Needs clarification" / "Under evaluation", thing that would generate a symbol liike `:"Needs Clarification"`. OFC it could be implemented using `.parameterize(separator: '_').to_sym`, but you would loose the control of how is defined
2. description field, it was meant to be used internally ( ex in admin, to define some kind of description on "what that state represents") 
3. system checkbox is meant flag the states designated as system ( not answered, evaluating, withdrawn etc), and it is a discriminator between core states and customization.
4. default checkbox was meant to designate the real "default state". While I am writing this, i realize i have not used it. ( I can descard it )  
5. I do understand they are special states. To be honest, I have added the states here to simplify the technical implementation. If i do not have them here, it would become a mess (filters, labels etc) to load some states from db and some other from code. I understand that is unforeseeable a status customization on "not answered".
6. "answerable" and "notifiable" are some flags that would allow you to enable or disable the notifications. And yes, they have been added in withdrawn context. 
7. regarding the Gamified option, this is added as some of the custom states that you may add can be considered "accepted", but the token / state could be different than "accepted"
8. I have added the CSS class instead of a simple color picker as the proposal states is represented differently inside the application. and usually you have 2 different colors in the badge ( the background and the text color ) 

From the things i answered to, i only see 1 or 2 things that i could improve: 
1. Remove the default checkbox. 
2. Remove the description field. (maybe)

alecslupu commented:
- [ ]  remove description field 
- [ ] remove from frontend the token attribute. 
- [ ] Replace any text "state" with "status" 
- [ ] Extract withdrawn to separate field
- [ ] Remove system 
- [ ] Remove gamified field 
- [ ] Remove notifiable field

carolromero reviewed (changes requested):
@alecslupu @andreslucena I could finally take a look at this. It's working great Alex, awesome work!

A couple of details that I'd change before merge:

- [x] Remove "token" column
- [ ] Change "Announcement title" to "Answer heading" (since Announcement can be confused with the actual component announcement).

On the other hand, we'll replace the CSS class for a color selector in another PR right?

alecslupu commented:
> @alecslupu @andreslucena I could finally take a look at this. It's working great Alex, awesome work!
> 
> A couple of details that I'd change before merge:
> 
>     * [x]  Remove "token" column
> 
>     * [x]  Change "Announcement title" to "Answer heading" (since Announcement can be confused with the actual component announcement).
> 
> 
> On the other hand, we'll replace the CSS class for a color selector in another PR right?

All the changes, required implemented. The Css selector, i would do it in a separte PR.

carolromero reviewed (dismissed):
> All the changes, required implemented. The Css selector, i would do it in a separte PR.

Nice! Checked and ready on my end. All yours @andreslucena

andreslucena reviewed (changes requested):
First round of suggestions, whenever you have time to check them out...

alecslupu commented on RELEASE_NOTES.md:
```diff
@@ -79,3 +79,19 @@ You need to change it to:
+
+For any of the above states you can customize the name, description, css class used by labels. You can also decide which states the user can receive a notification or an answer.
+
+You do not need to run any task to migrate the existing states, as we will automatically migrate the existing states to the new table.
```
To be honest, i would keep it here for the time being. There are a few community extensions that may be required to know about this change.

Apart of the change we did by moving the states to DB we also changed the scope naming. therefore, I would move to change the text as in suggestion.

andreslucena replied:
IMO as this is going to be highlighted in the https://decidim.org/blog releases notes, its a bit of a space waste to have it here too, but as we don't have too many points to communicate at the moment, I'll not block it. Although if we start to see more things for this release and we finish with 900 lines of Releases Notes, then we'll discuss about dropping this for sure. 

Can you review the writing? The description is for the first version and we need to update it to the current approach (i.e. not possible to customize the Withdrawn state and so on)

alecslupu commented on decidim-proposals/app/views/decidim/proposals/admin/proposal_states/_form.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="form__wrapper">
+  <div class="card pt-4">
+    <div class="card-section debate-fields">
```
```suggestion
    <div class="card-section">
```

alecslupu commented on decidim-proposals/app/views/decidim/proposals/admin/proposal_states/_form.html.erb:
```diff
@@ -0,0 +1,15 @@
+<div class="form__wrapper">
+  <div class="card pt-4">
+    <div class="card-section debate-fields">
+      <div class="row column hashtags__container">
```
```suggestion
      <div class="row column">
```

alecslupu commented on decidim-proposals/spec/system/private_space_proposal_spec.rb:
```diff
@@ -14,7 +14,7 @@
 
   let!(:participatory_space) { participatory_space_private }
 
-  let!(:component) { create(:component, manifest:, participatory_space:) }
```
fixed in a11f6f737b

andreslucena commented:
@alecslupu there's a merge conflict, can you check it out please?

alecslupu commented:
> @alecslupu there's a merge conflict, can you check it out please?

@andreslucena Conflict solved. This is ready for another round.

andreslucena reviewed (changes requested):
Two really small changes in the migrations and we're good to merge

alecslupu commented on decidim-proposals/db/migrate/20240110203502_add_state_id_to_decidim_proposals_proposals.rb:
```diff
@@ -0,0 +1,14 @@
+class AddStateIdToDecidimProposalsProposals < ActiveRecord::Migration[6.1]
+  def up
+    add_column :decidim_proposals_proposals, :decidim_proposals_proposal_state_id, :integer, index: true
+
```
Done in [d009a59](https://github.com/decidim/decidim/pull/12052/commits/d009a5960c5e0a3d7f3282a7044cede730361e11)

alecslupu commented on decidim-proposals/db/migrate/20240110203504_create_default_proposal_states.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+  end
+
+  def down; end
```
Done in [d009a59](https://github.com/decidim/decidim/pull/12052/commits/d009a5960c5e0a3d7f3282a7044cede730361e11)

andreslucena commented on decidim-elections/app/commands/decidim/elections/admin/import_proposals_to_elections.rb:
```diff
@@ -62,7 +62,7 @@ def create_answer_from_proposal(original_proposal)
         def proposals
           @proposals ||= if @form.import_all_accepted_proposals?
-                           Decidim::Proposals::Proposal.where(component: origin_component).where(state: "accepted")
+                           Decidim::Proposals::Proposal.where(component: origin_component).only_status(:accepted)
```
I think I'm not following you. You mean like having some kind of `define_method` with the status to bring a bit of syntax sugar to these scopes?

andreslucena replied:
Oh nevermind, we can ignore this one as it's from decidim-elections and this should be removed as that module was extracted to the `decidim-bulletin-board` repository. Can you remove this file please?

alecslupu replied:
Elections module has been removed. This is not relevant anymore.

andreslucena commented on decidim-budgets/app/commands/decidim/budgets/admin/import_proposals_to_budgets.rb:
```diff
@@ -75,8 +75,7 @@ def proposals
         def all_proposals
-          Decidim::Proposals::Proposal.where(component: origin_component)
-                                      .where(state: :accepted)
+          Decidim::Proposals::Proposal.where(component: origin_component).only_status(:accepted)
```
I think I'm not following you. You mean like having some kind of `define_method` with the status to bring a bit of syntax sugar to these scopes?

alecslupu replied:
Already applied in c46e30d7
